### PR TITLE
Minmax 3

### PR DESCRIPTION
**Pull request by Jiaqi-beep**
- Original NCL script: https://www.ncl.ucar.edu/Applications/Scripts/minmax_3.ncl
- Original NCL plot: https://www.ncl.ucar.edu/Applications/Images/minmax_3_lg.png
image

### CODE REVIEW

**Comment by Jiaqi-beep** [comment]
There's one text box missing on the plot. I'm looking into it.

File: Plots/Contours/NCL_minmax_3.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+import xarray as xr
+import matplotlib.pyplot as plt
+
+from geocat.viz import util as gvutil
+from geocat.viz import cmaps as gvcmaps
+
+################################################################
+# Definition of generate_2d_array and helper functions from https://github.com/NCAR/pyngl/blob/develop/src/ngl/__init__.py
+
+#  Globals for random number generator for generat_2d_array
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:29]:
```suggestion
#  Globals for random number generator for generate_2d_array
```

File: Plots/Contours/NCL_minmax_3.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+    r = dfran_rseq[dfran_iseq]
+    dfran_iseq = dfran_iseq + 1
+    return r
+
+def generate_2d_array(dims, num_low, num_high, minv, maxv, seed=0, \
+                      highs_at=None, lows_at=None):
+    """Generates a "nice" 2D array of pseudo random data, especially for use in
+    2D graphics.
+
+    dims -- a list (or array) containing the dimensions of the
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:61]: I think these should follow the formatting practices we've been using

File: Plots/Contours/NCL_minmax_3.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                pairs specifying where the highs will occur.  If this
+                argument appears, then its length must equal num_high and
+                the coordinates must be in the ranges specified in dims.
+    """
+
+    # Globals for random numbers
+    global dfran_iseq
+    dfran_iseq = seed
+
+    # Check arguments
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:87]: Not sure if we need these types of checks for this function, but if you want to put this function into comp, then checks would maybe be necessary.

File: Plots/Contours/NCL_minmax_3.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+
+from geocat.viz import util as gvutil
+from geocat.viz import cmaps as gvcmaps
+
+################################################################
+# Definition of generate_2d_array and helper functions from https://github.com/NCAR/pyngl/blob/develop/src/ngl/__init__.py
+
+#  Globals for random number generator for generat_2d_array
+dfran_iseq = 0
+dfran_rseq = [.749, .973, .666, .804, .081, .483, .919, .903,   \
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:31]: the '\' in this block can be taken out along with the trailing whitespace that was added in pyngl

File: Plots/Contours/NCL_minmax_3.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+
+###############################################################################
+# Helper function to add contour labels of local extrema with bounding boxes
+
+
+def plotLabels(coord_locations, label):
+    # Find contour value based on longitude and latitude coordinates
+    for coord in coord_locations:
+        # Note: second item of coord (lat) accesses the idex for the row number,
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:186]:
```suggestion
        # Note: second item of coord (lat) accesses the index for the row number,
```

File: Plots/Contours/NCL_minmax_3.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        # and first item of coord (lon) access the idex for the column number
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:187]:
```suggestion
        # and first item of coord (lon) access the index for the column number
```

File: Plots/Contours/NCL_minmax_3.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+import numpy as np
+import xarray as xr
+import matplotlib.pyplot as plt
+
+from geocat.viz import util as gvutil
+from geocat.viz import cmaps as gvcmaps
+
+################################################################
+# Definition of generate_2d_array and helper functions from https://github.com/NCAR/pyngl/blob/develop/src/ngl/__init__.py
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:27]:
```suggestion
# Definition of generate_2d_array and helper functions from
# https://github.com/NCAR/pyngl/blob/develop/src/ngl/__init__.py
```
Just for pep8 formatting sake

**Review by michaelavs — COMMENTED** [review]
Looks great overall, just some formatting things from the generate_2d_array function (basically pyngl formatting is different than our current formatting).

File: Plots/Contours/NCL_minmax_3.py (hunk 3 above)
Comment (Jiaqi-beep) [reply Plots/Contours/NCL_minmax_3.py]: Yep, I took them out!

File: Plots/Contours/NCL_minmax_3.py (hunk 2 above)
Comment (Jiaqi-beep) [reply Plots/Contours/NCL_minmax_3.py]: Updated.

File: Plots/Contours/NCL_minmax_3.py (hunk 4 above)
Comment (Jiaqi-beep) [reply Plots/Contours/NCL_minmax_3.py]: Updated. Lmk if this is what you had in mind!

File: Plots/Contours/NCL_minmax_3.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+def _dfran():
+    global dfran_iseq
+    global dfran_rseq
+    dfran_iseq = dfran_iseq % 100
+    r = dfran_rseq[dfran_iseq]
+    dfran_iseq = dfran_iseq + 1
+    return r
+
+def generate_2d_array(dims, num_low, num_high, minv, maxv, seed=0, \
+                      highs_at=None, lows_at=None):
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:55]:
```suggestion
```
I guess this isn't used in the function

File: Plots/Contours/NCL_minmax_3.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+#  Random number generator for generate_2d_array.
+def _dfran():
+    global dfran_iseq
+    global dfran_rseq
+    dfran_iseq = dfran_iseq % 100
+    r = dfran_rseq[dfran_iseq]
+    dfran_iseq = dfran_iseq + 1
+    return r
+
+def generate_2d_array(dims, num_low, num_high, minv, maxv, seed=0, \
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:54]:
```suggestion
def generate_2d_array(dims, num_low, num_high, minv, maxv, seed=0)
```
Update for comment on 55

File: Plots/Contours/NCL_minmax_3.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+        tmp_array[0, k] = 1. + (float(nx) - 1.) * _dfran()
+        tmp_array[1, k] = 1. + (float(ny) - 1.) * _dfran()
+        tmp_array[2, k] = -1.
+    for k in range(num_low, num_low + num_high):
+        # highs at random locations.
+        tmp_array[0, k] = 1. + (float(nx) - 1.) * _dfran()
+        tmp_array[1, k] = 1. + (float(ny) - 1.) * _dfran()
+        tmp_array[2, k] = 1.
+
+    # Initialize dmin and dmax to positive and negative inifity
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:110]:
```suggestion
    # Initialize dmin and dmax to positive and negative infinity
```

File: Plots/Contours/NCL_minmax_3.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    dmin = np.inf
+    dmax = -np.inf
+    # Initilize all values in out_array to average of minv and maxv
```
Comment (michaelavs) [inline Plots/Contours/NCL_minmax_3.py:113]:
```suggestion
    # Initialize all values in out_array to average of minv and maxv
```

**Review by michaelavs — COMMENTED** [review]
Just a couple more little updates to this and then it looks good to me!

**Review by michaelavs — APPROVED** [review]

**Comment by erogluorhan** [comment]
@anissa111 @michaelavs what is our plan with?

**Review by pilotchute — APPROVED** [review]

**Pull request merged**